src/main/java/hudson/remoting/AbstractByteBufferCommandTransport.java

```diff
@@ -0,0 +1,286 @@
+    /**
+     * The {@link #readState} when waiting for a frame header.
+     */
+    private static final int READ_STATE_NEED_HEADER = 0;
+    /**
+     * The {@link #readState} when waiting for the remaining byte of a frame header.
+     */
+    private static final int READ_STATE_MORE_HEADER = 1;
+    /**
+     * The {@link #readState} when waiting for {@link #readFrameRemaining} more bytes of frame body.
+     */
+    private static final int READ_STATE_FRAME_BODY = 2;
+    /**
+     * The {@link #readState} when a whole command has been read.
+     */
+    private static final int READ_STATE_COMMAND_READY = 3;
+    /**
+     * Our channel.
+     */
+    protected Channel channel;
+    /**
+     * The chunk header buffer.
+     */
+    private final ByteBuffer writeChunkHeader = ByteBuffer.allocate(2);
+    /**
+     * The transport frame size.
+     */
+    private int transportFrameSize = 8192;
+    /**
+     * The chunk body.
+     */
+    private ByteBuffer writeChunkBody = ByteBuffer.allocate(transportFrameSize);
+    /**
+     * The delegate, this is required as we cannot access some of the methods of {@link ChunkHeader} outside of the
+     * remoting module.
+     */
+    private CommandReceiver receiver;
+    /**
+     * The queue of data that has been received and is waiting to be processed.
+     */
+    private final ByteBufferQueue receiveQueue = new ByteBufferQueue(transportFrameSize);
+    /**
+     * Internal tracking of the read state. We use primitive field rather than an enum to enhance data locality
+     */
+    private int readState;
+    /**
+     * When {@link #readState} is {@link #READ_STATE_FRAME_BODY} this is the number of remaining bytes in the current
+     * frame.
+     */
+    private int readFrameRemaining;
+    /**
+     * When {@link #readState} is {@link #READ_STATE_MORE_HEADER} this is the first byte of the header,
+     * when {@link #READ_STATE_FRAME_BODY} this is the {@link ChunkHeader#parse(int, int)} parsed header.
+     */
+    private int readFrameHeader;
+    /**
+     * The offset in {@link #readCommandSizes} holding the current command length.
+     */
+    private int readCommandIndex = 0;
+    /**
+     * The command lengths read in so far. Normally we just use index 0 but there is the case where some commands
+     * may have arrived prior to {@link #setup(Channel, CommandReceiver)} in which case we just have to buffer all
+     * the commands we get and parse them later.
+     */
+    private int[] readCommandSizes = new int[16];
+    /**
+     * The queue used to stage output.
+     */
+    private ByteBufferQueue sendStaging = new ByteBufferQueue(transportFrameSize);
+
+    /**
+     * Write the packet.
+     *
+     * @param header the header to write.
+     * @param data   the data to write.
+     * @throws IOException if the data could not be written.
+     */
+    protected abstract void write(ByteBuffer header, ByteBuffer data) throws IOException;
```

Comment:
So InterruptedException is not an option here? The API is not symmetric then, and this command transport may be not so good for sync protocols